src/transaction/ada.ts

```diff
@@ -1,82 +1,218 @@
+  const txHash = hash_transaction(transactionBody);
+  const vKeyWitnesses = Vkeywitnesses.new();
+  for (const privateKey of privateKeys) {
+    makeWitness(privateKey, txHash, vKeyWitnesses)
+  }
+  const witnesses = TransactionWitnessSet.new();
+  witnesses.set_vkeys(vKeyWitnesses);
+  return Buffer.from(
+    Transaction.new(transactionBody, witnesses).to_bytes(),
+  ).toString('hex')
+};
+
+const addOutputs = (transactionBuilder: TransactionBuilder, tos: To[]) => {
+  let amount = 0
+  for (const to of tos) {
+    const value = new BigNumber(1000000).times(to.value).toString()
+    amount += Number(value)
+    addOutput(transactionBuilder, to.address, value)
+  }
+  return amount
+}
+
+const addInputs = async (transactionBuilder: TransactionBuilder, transferBtcBasedBlockchain: TransferBtcBasedBlockchain) => {
+  const { fromUTXO, fromAddress } = transferBtcBasedBlockchain
+  if (fromAddress) {
+    return addAddressInputs(transactionBuilder, fromAddress)
+  }
+  if (fromUTXO) {
+    return addUtxoInputs(transactionBuilder, fromUTXO)
+  }
+  throw new Error('Field fromAddress or fromUTXO is not filled.')
+}
+
+const addAddressInputs = async (transactionBuilder: TransactionBuilder, fromAddresses: FromAddress[]) => {
+  let amount = 0
+  const privateKeysToSign: string[] = [];
+  for (const fromAddress of fromAddresses) {
+    const { address, privateKey } = fromAddress
+    const utxos: AdaUtxo[] = await adaGetUtxos(address)
+    for (const utxo of utxos) {
+      amount += parseInt(utxo.value);
+      addInput(transactionBuilder, privateKey, utxo)
+      privateKeysToSign.push(fromAddress.signatureId || fromAddress.privateKey)
+    }
+  }
+  return { amount, privateKeysToSign }
+}
+
+const addUtxoInputs = async (transactionBuilder: TransactionBuilder, fromUTXOs: FromUTXO[]) => {
+  let amount = 0
+  const privateKeysToSign: string[] = [];
+  for (const utxo of fromUTXOs) {
+    const transaction = await adaGetTransaction(utxo.txHash)
+    const output = transaction.outputs.find(output => output.index === utxo.index)
+    if (output) {
+      const value = output.value;
+      amount += parseInt(value);
+      addInput(transactionBuilder, utxo.privateKey, { value, ...utxo })
+      privateKeysToSign.push(utxo.signatureId || utxo.privateKey)
+    }
   }
+  return { amount, privateKeysToSign }
+}
 
-  const prvKey = CardanoWasm.Bip32PrivateKey.from_128_xprv(
-      Buffer.from(body.privateKey, 'hex')
-    ).to_raw_key();
+const addOutput = (transactionBuilder: TransactionBuilder, address: string, amount: string) => {
+  transactionBuilder.add_output(TransactionOutput.new(
+    Address.from_bech32(address),
+    Value.new(BigNum.from_str(amount)),
+  ));
+}
+
+const addInput = (transactionBuilder: TransactionBuilder, privateKey: string, utxo: AdaUtxo) => {
+  const privateKeyCardano = Bip32PrivateKey.from_128_xprv(
+    Buffer.from(privateKey, 'hex'),
```

Comment:
@samuelsramko Why I need a here private key? In the case of BTC, this doesn't need the private key. With this setup, I am not able to create a transaction and sign it later. Please propose some workaround.